lib/auth/scopes/accesscontrol/service.go

```diff
@@ -0,0 +1,194 @@
+// Teleport
+// Copyright (C) 2025 Gravitational, Inc.
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+package accesscontrol
+
+import (
+	"context"
+	"log/slog"
+
+	"github.com/gravitational/trace"
+
+	"github.com/gravitational/teleport"
+	scopedrolev1 "github.com/gravitational/teleport/api/gen/proto/go/teleport/scopedrole/v1"
+	"github.com/gravitational/teleport/api/types"
+	"github.com/gravitational/teleport/lib/authz"
+)
+
+// Config contains the parameters for [New].
+type Config struct {
+	Authorizer authz.Authorizer
+	Logger     *slog.Logger
+}
+
+// Server is the [scopedrolev1.ScopedRoleServiceServer] returned by [New].
+type Server struct {
+	scopedrolev1.UnsafeScopedRoleServiceServer
+
+	authorizer authz.Authorizer
+	logger     *slog.Logger
+}
+
+// New returns the auth server implementation for the scoped access control
+// service, including the gRPC interface, authz enforcement, and business logic.
+func New(c Config) (*Server, error) {
+	if c.Authorizer == nil {
+		return nil, trace.BadParameter("missing Authorizer")
+	}
+	if c.Logger == nil {
+		c.Logger = slog.With(teleport.ComponentKey, "scopes")
+	}
+
+	return &Server{
+		authorizer: c.Authorizer,
+		logger:     c.Logger,
+	}, nil
+}
+
+// CreateScopedRole implements [scopedrolev1.ScopedRoleServiceServer].
+func (s *Server) CreateScopedRole(ctx context.Context, req *scopedrolev1.CreateScopedRoleRequest) (*scopedrolev1.CreateScopedRoleResponse, error) {
+	authzContext, err := s.authorizer.Authorize(ctx)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	if !authz.HasBuiltinRole(*authzContext, string(types.RoleAdmin)) {
+		s.logger.WarnContext(ctx, "user does not have permission to create scoped roles", "user", authzContext.User.GetName())
+		return nil, trace.AccessDenied("user %q does not have permission to create scoped roles", authzContext.User.GetName())
+	}
+
+	return (scopedrolev1.UnimplementedScopedRoleServiceServer{}).CreateScopedRole(ctx, req)
```

Comment:
Can we replace this with a reference to a single field that will now be equal to `scopedrolev1.UnimplementedScopedRoleServiceServer{}` but will be replaced by actual implementation in the future?

Here and in all other methods.

Reply:
Fair enough.